File: apps/android/app/src/test/java/com/tonezen/app/domain/progress/CycleListenProgressTest.kt
package com.tonezen.app.domain.progress

import com.tonezen.app.domain.model.AudiobookProgress
import com.tonezen.app.domain.model.Book
import com.tonezen.app.domain.model.ContentType
import com.tonezen.app.domain.model.Cycle
import com.tonezen.app.domain.model.Track
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class CycleListenProgressTest {
    private val bookOne = book("book-1", "book-one", 0)
    private val bookTwo = book("book-2", "book-two", 1)
    private val cycle = Cycle(
        id = "cycle-1",
        slug = "cycle",
        title = "Cycle",
        bookOrder = listOf(bookOne.slug, bookTwo.slug),
        books = listOf(bookOne, bookTwo),
    )
    private val tracksByBookId = mapOf(
        bookOne.id to listOf(track("t1", bookOne.id, 0, 100_000)),
        bookTwo.id to listOf(
            track("t2", bookTwo.id, 0, 100_000),
            track("t3", bookTwo.id, 1, 100_000),
        ),
    )

    @Test
    fun `cycle fraction combines progress across books`() {
        val fraction = resolveCycleListenFraction(
            cycle = cycle,
            tracksByBookId = tracksByBookId,
            progressByBookId = mapOf(
                bookOne.id to progress(bookOne.id, "t1", 100_000),
                bookTwo.id to progress(bookTwo.id, "t2", 50_000),
            ),
        )
        assertEquals(0.5f, fraction)
    }

    @Test
    fun `resume returns first track when nothing listened`() {
        val resume = resolveCycleResumeTarget(cycle, tracksByBookId, emptyMap())
        assertEquals("t1", resume?.track?.id)
        assertEquals(0L, resume?.startPositionMs)
    }

    @Test
    fun `resume continues partial chapter`() {
        val resume = resolveCycleResumeTarget(
            cycle,
            tracksByBookId,
            mapOf(
                bookOne.id to progress(bookOne.id, "t1", 100_000),
                bookTwo.id to progress(bookTwo.id, "t2", 40_000),
            ),
        )
        assertEquals("t2", resume?.track?.id)
        assertEquals(40_000L, resume?.startPositionMs)
    }

    @Test
    fun `resume starts next chapter after completed one`() {
        val resume = resolveCycleResumeTarget(
            cycle,
            tracksByBookId,
            mapOf(
                bookOne.id to progress(bookOne.id, "t1", 100_000),
                bookTwo.id to progress(bookTwo.id, "t2", 100_000),
            ),
        )
        assertEquals("t3", resume?.track?.id)
        assertEquals(0L, resume?.startPositionMs)
    }

    @Test
    fun `cycle fraction is null without durations`() {
        val fraction = resolveCycleListenFraction(
            cycle = cycle,
            tracksByBookId = mapOf(bookOne.id to listOf(track("t1", bookOne.id, 0, null))),
            progressByBookId = emptyMap(),
        )
        assertNull(fraction)
    }

    @Test
    fun `continue state returns latest resumable chapter in cycle`() {
        val continueState = resolveCycleContinueState(
            cycle = cycle,
            tracksByBookId = tracksByBookId,
            progressByBookId = mapOf(
                bookOne.id to progress(bookOne.id, "t1", 95_000),
                bookTwo.id to progress(bookTwo.id, "t2", 12_000, updatedAtEpochMs = 2L),
            ),
        )
        assertEquals(BookContinueState("t2", 12_000), continueState)
    }

    @Test
    fun `continue state is null for another cycle without its own progress`() {
        val otherCycle = Cycle(
            id = "cycle-2",
            slug = "cycle-2",
            title = "Other",
            bookOrder = listOf("book-three"),
            books = listOf(
                book("book-3", "book-three", 0),
            ),
        )
        val otherTracks = mapOf(
            "book-3" to listOf(track("t3", "book-3", 0, 100_000)),
        )
        assertNull(
            resolveCycleContinueState(
                otherCycle,
                otherTracks,
                mapOf(bookTwo.id to progress(bookTwo.id, "t2", 12_000)),
            ),
        )
    }

    private fun book(id: String, slug: String, sortOrder: Int) = Book(
        id = id,
        slug = slug,
        contentType = ContentType.AUDIOBOOK,
        title = slug,
        author = "Author",
    )

    private fun track(id: String, bookId: String, sortOrder: Int, durationMs: Long?) = Track(
        id = id,
        bookId = bookId,
        sortOrder = sortOrder,
        title = id,
        filename = "$id.mp3",
        durationMs = durationMs,
        localPath = null,
    )

    private fun progress(bookId: String, trackId: String, positionMs: Long, updatedAtEpochMs: Long = 1L) =
        AudiobookProgress(
            bookId = bookId,
            trackId = trackId,
            positionMs = positionMs,
            updatedAtEpochMs = updatedAtEpochMs,
        )
}

File: apps/desktop/tests/cycleListenProgress.test.ts
import { describe, expect, it } from "vitest";
import {
  orderedCycleEntriesFromResume,
  resolveBookListenedMs,
  resolveCycleResumeTarget,
} from "../src/shared/cycleListenProgress.js";
import type { AudiobookProgress, Book, Cycle, Track } from "../src/shared/types.js";

const tracks: Track[] = [
  { id: "t1", bookId: "b1", sortOrder: 0, title: "One", filename: "1.mp3", durationMs: 100_000 },
  { id: "t2", bookId: "b1", sortOrder: 1, title: "Two", filename: "2.mp3", durationMs: 100_000 },
];

describe("resolveBookListenedMs", () => {
  it("sums completed chapters and current position", () => {
    const progress: AudiobookProgress = {
      bookId: "b1",
      trackId: "t2",
      positionMs: 40_000,
      updatedAt: "2026-01-01T00:00:00.000Z",
    };
    expect(resolveBookListenedMs(tracks, progress)).toBe(140_000);
  });

  it("returns zero when track does not belong to the book", () => {
    const progress: AudiobookProgress = {
      bookId: "b1",
      trackId: "t1",
      positionMs: 10_000,
      updatedAt: "2026-01-01T00:00:00.000Z",
    };
    expect(resolveBookListenedMs([{ ...tracks[0], bookId: "other" }], progress)).toBe(0);
  });
});

const bookOne: Book = {
  id: "book-1",
  slug: "book-one",
  title: "Book one",
  contentType: "audiobook",
};
const bookTwo: Book = {
  id: "book-2",
  slug: "book-two",
  title: "Book two",
  contentType: "audiobook",
};
const cycle: Cycle = {
  id: "cycle-1",
  slug: "cycle",
  title: "Cycle",
  bookOrder: [bookOne.slug, bookTwo.slug],
  books: [bookOne, bookTwo],
};
const cycleTracks = new Map<string, Track[]>([
  [bookOne.id, [{ id: "t1", bookId: bookOne.id, sortOrder: 0, title: "t1", filename: "1.mp3", durationMs: 100_000 }]],
  [
    bookTwo.id,
    [
      { id: "t2", bookId: bookTwo.id, sortOrder: 0, title: "t2", filename: "2.mp3", durationMs: 100_000 },
      { id: "t3", bookId: bookTwo.id, sortOrder: 1, title: "t3", filename: "3.mp3", durationMs: 100_000 },
    ],
  ],
]);

describe("resolveCycleResumeTarget", () => {
  it("returns first track when nothing listened", () => {
    const resume = resolveCycleResumeTarget(cycle, cycleTracks, new Map());
    expect(resume?.track.id).toBe("t1");
    expect(resume?.startPositionMs).toBe(0);
  });

  it("continues partial chapter", () => {
    const progress = new Map<string, AudiobookProgress>([
      [bookOne.id, { bookId: bookOne.id, trackId: "t1", positionMs: 100_000, updatedAt: "2026-01-01T00:00:00.000Z" }],
      [bookTwo.id, { bookId: bookTwo.id, trackId: "t2", positionMs: 40_000, updatedAt: "2026-01-01T00:00:00.000Z" }],
    ]);
    const resume = resolveCycleResumeTarget(cycle, cycleTracks, progress);
    expect(resume?.track.id).toBe("t2");
    expect(resume?.startPositionMs).toBe(40_000);
  });
});

describe("orderedCycleEntriesFromResume", () => {
  it("orders entries from resume target forward", () => {
    const resume = {
      book: bookTwo,
      track: cycleTracks.get(bookTwo.id)![1],
      startPositionMs: 0,
    };
    const entries = orderedCycleEntriesFromResume(cycle, cycleTracks, resume);
    expect(entries.map((entry) => entry.track.id)).toEqual(["t3"]);
  });
});
